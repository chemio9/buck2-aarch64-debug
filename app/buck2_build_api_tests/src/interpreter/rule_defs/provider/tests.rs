/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under both the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree and the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree.
 */

use buck2_build_api::interpreter::rule_defs::provider::callable::register_provider;
use buck2_build_api::interpreter::rule_defs::register_rule_defs;
use buck2_core::bzl::ImportPath;
use buck2_interpreter_for_build::interpreter::testing::Tester;
use indoc::indoc;

fn provider_tester() -> Tester {
    let mut tester = Tester::new().unwrap();
    tester.additional_globals(register_rule_defs);
    tester.additional_globals(register_provider);
    tester
}

#[test]
fn creates_providers() -> anyhow::Result<()> {
    // TODO(nmj): Starlark doesn't let you call 'new_invoker()' on is_mutable types.
    //                 Once that's fixed, make sure we can call 'FooInfo' before the module is
    //                 frozen.
    let mut tester = provider_tester();

    tester.run_starlark_test(indoc!(
        r#"
    FooInfo = provider(fields=["bar", "baz"])
    FooInfo2 = FooInfo
    #frozen_foo_1 = FooInfo(bar="bar_f1", baz="baz_f1")
    #frozen_foo_2 = FooInfo(bar="bar_f2")

    assert_eq('provider(fields={"f1": provider_field(typing.Any, default=None)})', repr(provider(fields=["f1"])))
    assert_eq('provider[FooInfo](fields={"bar": provider_field(typing.Any, default=None), "baz": provider_field(typing.Any, default=None)})', repr(FooInfo))
    assert_eq('provider[FooInfo](fields={"bar": provider_field(typing.Any, default=None), "baz": provider_field(typing.Any, default=None)})', repr(FooInfo2))

    def test():
        assert_eq('provider[FooInfo](fields={"bar": provider_field(typing.Any, default=None), "baz": provider_field(typing.Any, default=None)})', repr(FooInfo))
        assert_eq('provider[FooInfo](fields={"bar": provider_field(typing.Any, default=None), "baz": provider_field(typing.Any, default=None)})', repr(FooInfo2))

        #assert_eq("FooInfo(bar=\"bar_f1\", baz=\"baz_f1\")", repr(frozen_foo1))
        #assert_eq("bar_f1", frozen_foo1.bar)
        #assert_eq("baz_f1", frozen_foo1.baz)
        #assert_eq("FooInfo(bar=\"bar_f2\", baz=None)", repr(frozen_foo2))
        #assert_eq("bar_f2", frozen_foo2.bar)
        #assert_eq(None, frozen_foo2.baz)

        foo_1 = FooInfo(bar="bar_1", baz="baz_1")
        foo_2 = FooInfo(bar="bar_2")

        assert_eq('provider[FooInfo](fields={"bar": provider_field(typing.Any, default=None), "baz": provider_field(typing.Any, default=None)})', repr(FooInfo))
        assert_eq("FooInfo(bar=\"bar_1\", baz=\"baz_1\")", repr(foo_1))
        assert_eq("bar_1", foo_1.bar)
        assert_eq("baz_1", foo_1.baz)
        assert_eq("FooInfo(bar=\"bar_2\", baz=None)", repr(foo_2))
        assert_eq("bar_2", foo_2.bar)
        assert_eq(None, foo_2.baz)

        assert_eq("{\"bar\":\"bar_1\",\"baz\":\"baz_1\"}", json.encode(foo_1))
    "#
    ))?;

    tester.run_starlark_test_expecting_error(
        indoc!(
            r#"
    FooInfo = provider(fields=["bar", "baz"])

    def test_compile_time():
        foo_1 = FooInfo(bar="bar1")
        foo_1.quz

    def test():
        pass
    "#
        ),
        "The attribute `quz` is not available on the type `FooInfo`",
    );

    tester.run_starlark_test_expecting_error(
        indoc!(
            r#"
    list = []
    list.append(provider(fields=["bar", "baz"]))
    "#
        ),
        "must be assigned to a variable",
    );

    // Make sure that frozen UserProvider instances work
    let mut tester = provider_tester();
    tester.add_import(
        &ImportPath::testing_new("root//provider:def1.bzl"),
        indoc!(
            r#"
            FooInfo = provider(fields=["foo"])
            "#
        ),
    )?;
    tester.add_import(
        &ImportPath::testing_new("root//provider:def2.bzl"),
        indoc!(
            r#"
            load("//provider:def1.bzl", "FooInfo")
            foo = FooInfo(foo="foo1")
            "#
        ),
    )?;
    tester.run_starlark_test(indoc!(
        r#"
        load("//provider:def2.bzl", "foo")
        def test():
            assert_eq('FooInfo(foo="foo1")', repr(foo))
        "#
    ))?;

    Ok(())
}

#[test]
fn test_builtin_provider_as_type() {
    let mut tester = provider_tester();
    tester
        .run_starlark_bzl_test(indoc!(
            r#"
        def test():
            configuration_info = ConfigurationInfo(constraints={}, values={})
            assert_true(isinstance(configuration_info, ConfigurationInfo))
            assert_true(isinstance(configuration_info, Provider))
            assert_true(not isinstance(configuration_info, DefaultInfo))
        "#
        ))
        .unwrap();
}

#[test]
fn test_user_defined_provider_as_type() {
    let mut tester = provider_tester();
    tester
        .run_starlark_bzl_test(indoc!(
            r#"
            FooInfo = provider(fields=["foo"])
            def test():
                foo = FooInfo(foo="foo1")
                assert_true(isinstance(foo, FooInfo))
                assert_true(isinstance(foo, Provider))
                assert_true(not isinstance(foo, DefaultInfo))
            "#
        ))
        .unwrap();
}

#[test]
fn test_user_defined_provider_creator_typecheck() {
    let mut tester = provider_tester();
    tester.run_starlark_bzl_test_expecting_error(
        indoc!(
            r#"
                Pkg = provider(fields = ["x"])

                def create_pkg():
                    return Pkg(y = 1)

                def test():
                    pass
    "#
        ),
        "Unexpected parameter named `y`",
    );
}

#[test]
fn test_provider_non_unique_fields() {
    let mut tester = provider_tester();
    tester.run_starlark_bzl_test_expecting_error(
        indoc!(
            r#"
                Pkg = provider(fields = ["x", "x"])

                def test():
                    pass
            "#
        ),
        "non-unique field names",
    )
}
