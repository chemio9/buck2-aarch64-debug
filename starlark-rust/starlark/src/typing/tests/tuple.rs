/*
 * Copyright 2019 The Starlark in Rust Authors.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use crate::typing::tests::TypeCheck;

#[test]
fn test_tuple() {
    TypeCheck::new().check(
        "tuple",
        r#"
def empty_tuple_fixed_name() -> (): return tuple()
def empty_tuple_name_fixed() -> tuple: return ()
"#,
    );
}

#[test]
fn test_tuple_ellipsis() {
    TypeCheck::new().check(
        "tuple_ellipsis",
        r#"
def f(t: tuple[int, ...]) -> int:
    return t[0]

def g():
    # Good.
    f((1, 2, 3))

    # Bad.
    f((1, "x"))
"#,
    );
}
