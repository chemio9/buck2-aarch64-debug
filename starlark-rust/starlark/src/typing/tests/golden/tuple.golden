# @generated
# To regenerate, run:
# ```
# STARLARK_RUST_REGENERATE_GOLDEN_TESTS=1 cargo test -p starlark --lib
# ```

Code:
def empty_tuple_fixed_name() -> (): return tuple()
def empty_tuple_name_fixed() -> tuple: return ()

No errors.

Compiler typechecker (eval):
No errors.
